### DISCUSSION TITLE
black_hole 插件有疑问

### DISCUSSION THREAD

**Discussion by sbwml** [IrineSistiana/mosdns#717] [Q&A] [answered]
近日我在 openwrt mosdns luci 插件中有使用到 `black_hole` 插件进行自定义应答，在 wiki 中有看到它支持 `black_hole 1.1.1.1 2.2.2.2 ...` 多 ip 的形式，请问定义多个 ip 后它是有意义的吗？因为在测试中，它似乎永远返回 `1.1.1.1` 也就是第一个。

---------------

![image](https://github.com/IrineSistiana/mosdns/assets/16485166/c402b635-37c2-4b23-98e1-f68d74bfe36c)

![image](https://github.com/IrineSistiana/mosdns/assets/16485166/68337f53-3696-44f5-af07-d98aa70999b7)

如图，我为某个 IP 池使用 `black_hole` 定义了三个IP，在 dns 解析测试中，它确实得到了我定义的那三个IP 地址。
但是在浏览器、curl、wget、ping 等各种客户端上，它在任何条件下都是使用第一个 IP。

------------

我的疑问是，有办法让 `black_hole 1.1.1.1 2.2.2.2 3.3.3.3 ...` 多个ip的情况下随机打乱的方式响应吗？目的为了客户端能真实使用到这些地址，而不是永远只有 `1.1.1.1`

**Comment by sbwml** [comment]
cc @IrineSistiana

**Comment by IrineSistiana** [comment]
black_hole 没有问题。它会返回多个IP。如何处理/利用多个IP是客户端的事。

**Comment by sbwml** [comment]
好吧，我自己琢磨琢磨

**Comment by sbwml** [comment, marked as answer]
问题解决了，我的方法是对 black_hole 插件做洗牌算法 https://github.com/sbwml/luci-app-mosdns/commit/b8b7d64ba9bcf499ad228bfc1b9c2342d7d49a96 ，通过打乱地址的方式响应给客户端，这样客户端就可以在 `black_hole 1.1.1.1 2.2.2.2 ... ...` 范围中随机得到它的ip顺序。而不是固定只拿到 1.1.1.1

我在能测试的客户端上都测试了，他们都是只使用响应的第一个ip；我还测试了几个cdn 域名，第一次解析后得到了5个ip，过了ttl 时间后，再次解析还是得到那五个ip，但是他们的顺序也是改变了的


改变它的目的是让客户端可以随机使用到被 black_hole 定义的全部 ip，而不是只有第一个ip 可以被使用。因为手机、电脑你给他响应
```
xxx.com. 60s A 1.0.0.0
xxx.com. 60s A 2.0.0.0
xxx.com. 60s A 3.0.0.0
```

它真的是永远只会使用 1.0.0.0。

**Reply by IrineSistiana** [reply]
我大概明白你的意思了，你想给 cf 负载均衡。但我觉得 dns 层面负载均衡对 cf 没意义。 因为 cf 不靠 dns 负载均衡 (目测是靠任播+数据中心内分载)。

另外，这个 patch 有 bug ，存在竞争。

**Reply by sbwml** [reply]
我是想通过这个东西 https://github.com/XIU2/CloudflareSpeedTest 得到比较优质的地址，在局域网内的设备在访问套了 cf 的站点时，可以用前面测试出来的节点去访问，这样局域网下的设备访问cf cdn 的站点时可以得到比较友好的速度。因为套了cf 的一些网站，在不开科学上网的时候，往往比较堪忧。

草根选手不懂些，求大佬帮帮给改改

**Reply by IrineSistiana** [reply]
没必要在 black hole 洗牌。因为不会起到负载均衡的效果。还有，如果你配置了 cache，命中 cache 会自动触发洗牌。

所以什么都不用改。

**Comment by qingtian110** [comment]
> 我大概明白你的意思了，你想给 cf 负载均衡。但我觉得 dns 层面负载均衡对 cf 没意义。 因为 cf 不靠 dns 负载均衡 (目测是靠任播+数据中心内分载)。

有意义的，而且意义巨大。他这个不是为了负载均衡的，而是 IP 优选。我家的网络，用了优选后的 IP，去使用 cloudflare worker ，我多线程下载文件的速度直接飙到 77M/s 去了。如果没有做优选IP的话，多线程速度 5M/s 每秒极限了

**Reply by qingtian110** [reply]
另外cf 的ip 超级古怪，像这个 104.16.255.180 ip，我家有两条宽带，电信和联通，联通去用这个ip 速度超快的，但是电信用它就蜗牛速度了，用 black_hole 给各自的路由设置优选ip 后，两条宽带访问 Cloudflare 的站点速度都变得很快了，他还有个牛逼的地方    是所有使用了 Cloudflare 的域名     都会变得很快

**Reply by IrineSistiana** [reply]
我知道 cf 优选。我的意思是优选出来的 IP 没必要每次返回都随机洗牌。

**Reply by qingtian110** [reply]
洗牌好像是有用的，我对比了加了洗牌的二进制和没有洗牌的二进制。它出现了两种情况

我用了两个假IP 和 一个真实 CF IP 来测试

`black_hole 1.2.0.0 1.2.0.1 104.16.255.180`

没有洗牌：

```
ping -4 gh.api.99988866.xyz
```
尝试了 5 次，得到的结果都是 `PING gh.api.99988866.xyz (1.2.0.0) 56(84) bytes of data.`，同时浏览器无痕模式所有 cf 加速的页面都打不开


加了洗牌的：
```
ping -4 gh.api.99988866.xyz
```
第一次结果是：`PING gh.api.99988866.xyz (1.2.0.1) 56(84) bytes of data.`
第二次结果是：
```
PING gh.api.99988866.xyz (104.16.255.180) 56(84) bytes of data.
64 bytes from 104.16.255.180 (104.16.255.180): icmp_seq=1 ttl=53 time=153 ms
64 bytes from 104.16.255.180 (104.16.255.180): icmp_seq=2 ttl=53 time=153 ms
64 bytes from 104.16.255.180 (104.16.255.180): icmp_seq=3 ttl=53 time=150 ms
64 bytes from 104.16.255.180 (104.16.255.180): icmp_seq=4 ttl=53 time=151 ms
```

----------

得到的结论

没有洗牌：
`black_hole 1.2.0.0 1.2.0.1 104.16.255.180` = `black_hole 1.2.0.0` ，无论后面加多少个IP，它最终都是 `black_hole 1.2.0.0` 的效果。

加了洗牌：`black_hole 1.2.0.0 1.2.0.1 104.16.255.180`  black_hole 的三个 IP 都有在浏览器和 ping 出现到，他设置的多个 IP 是有作用的。


在开启cache 插件和关闭 cache 插件测试，得到的结果是一致的

**Reply by qingtian110** [reply]
在用 CloudflareST 找出来的优选IP，他通常会有五六个速度是非常快的，在加了洗牌的二进制上，这加上去的五六个ip 都会被随机使用到，不加洗牌的二进制，添加6个优选IP 上去，有五个是纯摆设状态，因为只有第一个IP 会被浏览器拿到

**Reply by IrineSistiana** [reply]
> 在加了洗牌的二进制上，这加上去的五六个ip 都会被随机使用到，不加洗牌的二进制，添加6个优选IP 上去，有五个是纯摆设状态，因为只有第一个IP 会被浏览器拿到

洗牌 = 五六个ip 都会被随机使用 = dns层面/IP层面负载均衡 = 对于cf没用

而且算平均速度的话，你填很多个ip平均速度会被拉低，还不如只填一个最快的IP。

**Comment by daiaji** [comment]
负载均衡难说，但我猜域名解析出多个A记录用处主要是冗余，万一挂了一个ip后还能用别的。
https://zh.m.wikipedia.org/zh-cn/DNS%E8%BD%AE%E5%BE%AA
看起来确实是容错用。
